products/views.py

```diff
@@ -1,3 +1,68 @@
-from django.shortcuts import render
+from django.views     import View
+from django.http      import JsonResponse
+from django.db.models import Q
 
-# Create your views here.
+from products.models import *
+from reviews.models  import *
+
+class ProductListView(View):
+    def get(self,request):
+        country_id   = request.GET.getlist('country_id', None)
+        winery_id    = request.GET.getlist('winery_id', None)
+        grape        = request.GET.getlist('grape', None)
+        type_id      = request.GET.getlist('type_id', None)
+        pairing_id   = request.GET.getlist('pairing_id', None)
+        min_price    = request.GET.get('min_price', 0)
+        max_price    = request.GET.get('max_price', 500000)
+        offset       = int(request.GET.get('offset' , 0))
+        limit        = int(request.GET.get('limit' , 100))
+                
+        q = Q(price__range = (min_price, max_price))
+        
+        if country_id:
+            q &= Q(winery_id__country_id__in = country_id)
+            
+        if winery_id:
+            q &= Q(winery_id__in = winery_id)
+
+        if grape:
+            q &= Q(grape__in = grape)
+            
+        if type_id:
+            q &= Q(type_id__in = type_id)
+            
+        if pairing_id:
+            q &= Q(pairings__id__in = pairing_id)
+
+        products = Product.objects.filter(q)[offset:offset+limit]
+
+        result = [{
+            'id'       : product.id,
+            'name'     : product.name,
+            'price'    : product.price,
+            'grape'    : product.grape,
+            'bold'     : product.bold,
+            'tannic'   : product.tannic,
+            'sweet'    : product.sweet,
+            'acidic'   : product.acidic,
+            'country'  : {
+                'id'  : product.winery.country.id,
+                'name': product.winery.country.name
+            },
+            'winery'   : {
+                'id'  : product.winery.id,
+                'name': product.winery.name
+            },
+            'type'     : {
+                'id'  : product.type.id,
+                'name': product.type.name
+            },
+            'image_url': product.image_url,
+            'pairing'  : [productpairing.pairing.name for productpairing in product.productpairings.all()],
+            'review'   : {
+                'rating_count' : Review.objects.filter(product = product).count(),
```

Comment:
이부분은 line 37 코드에 `annotate()` method를 사용해서 리펙토링해보세요~
```suggestion
                'rating_count' : Review.objects.filter(product = product).count(),
```

Reply:
annotate를 활용하여 변경완료했습니다.